### PR TITLE
[opt](nereids) adjust cost model for BroadCastJoin and PartitionJoin

### PR DESCRIPTION
**Pull request by englefly**
## Proposed changes
we add penalty for broadcast join (bc for brief in the following).
the intuition of penalty is as follow:
1. if the build side is very small (< 1M), we prefer bc, and set `penalty=1`, which means no penalty
2. if build side is more than 1M, we consider the ratio of the probe row count to the build row count. the less the ratio is, the higher penalty is.

this pr has positive impact on tpch queries. Only q3 is changed. in out test (tpch 1T, 3BE) q3 improved from 5.1sec to 2.5 sec.
this pr has positive impact on tpcds queries. test on tpcds sf100 (3BE), cold run improve from 163 sec to 156 sec, hot run improves from 155 sec to 149 sec

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java [hunk 1]
```diff
@@ -252,10 +271,19 @@ public Cost visitPhysicalHashJoin(
                     leftRowCount + rightRowCount,
                     penalty);
         }
+
+        if (context.isBroadcastJoin()) {
+            double broadcastJoinPenalty = broadCastJoinBalancePenalty(probeStats, buildStats);
+            return CostV1.of(leftRowCount * broadcastJoinPenalty + rightRowCount + outputRowCount,
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:277]: I think we would better to set up cost system on more dimension, for distinguishing the bc and shuffle, for example, we would consider exchange data cost, parallelism info to compare these two, by simulating the real execution difference between bc and shuffle as close as possible. In comparison, the introduced penalty parameter is meaningless, such as BROADCAST_JOIN_SKEW_RATIO, and will easily to be broken under different scenario in the future.

Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:277]: this idea can be implemented in CostModelV2. In V1, we do not have this dimension

**Review by xzj7019 — APPROVED** [review]

**Comment by englefly** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**